Blocks: Add generic on save modal for site editor and post editor to go back to Launchpad

sharpirate opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes [LaunchPad - Site Editor Save Modal - Adapt for non-LIB flows](https://github.com/Automattic/wp-calypso/issues/71717) and [LaunchPad - adapt and enable the back-to-launchpad modal in the post/page editor](https://github.com/Automattic/wp-calypso/issues/72093)

## Proposed changes:
This PR provides a generic modal for all Launchpad flows to allow the user to go back to the Launchpad screen from the Site Editor and Post Editor once they have saved their progress.

<img width="866" alt="image" src="https://user-images.githubusercontent.com/20927667/213217851-efc777e5-b74a-4b00-8219-75dea44cd11c.png">

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
paYE8P-1Kg-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:

### Testing on WPCOM
1. Sandbox `public-api.wordpress.com` along with a Launchpad enabled site
2. Run `bin/jetpack-downloader test jetpack add/launchpad_modal_all_flows` on your sandbox
3. Go to the Site Editor
4. Make a change and save. You should see the modal appear.
5. Go to the Post Editor.
6. Test publishing a new post and updating the post once it's published. You should see the modal appear on save.
7. Test with different flows - `newsletter`, `link-in-bio`, `free`, `videopress`.

sharpirate committed:
[not verified] add support for all flows in the site editor on save modal

sharpirate committed:
[not verified] changelog

jeherve reviewed (dismissed):
Tests well for me 🚢 

<img width="579" alt="Screenshot 2023-01-16 at 12 33 47" src="https://user-images.githubusercontent.com/426388/212668955-dd133577-f319-415f-8ec8-8be98199019a.png">
<img width="738" alt="Screenshot 2023-01-16 at 12 28 50" src="https://user-images.githubusercontent.com/426388/212668962-1645d00c-ff05-46fc-9453-0dc5a6741118.png">

sharpirate committed:
[not verified] add modal to post editor

sharpirate commented:
Hi @jeherve, thanks for testing that! After a discussion inside cylon, we decided to have one generic version of the modal for all flows. I've also added support for the post editor. Feel free to review / test again.

jeherve reviewed (commented):
Should the modal appear even before the post is published? It seems to be happening right now, and I personally find it a bit annoying when I'm writing a draft.

<img width="1247" alt="image" src="https://user-images.githubusercontent.com/426388/213256527-510efaed-1b9f-4336-8413-4d0c5af179f3.png">

Maybe we could wait until the post has been published?

sharpirate commented:
@jeherve, I'm not seeing that:

https://user-images.githubusercontent.com/20927667/213273925-35d9b0f6-c85e-4740-93ab-80da83086093.mov

jeyip commented:
## Testing
### Requirements

#### Link In Bio Flow
**Site editor modal** 
- [x] Modal appears when editing or saving site modifications, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

**Post editor modal** 
- [x] Modal appears when updating or saving the post, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

#### Newsletter
**Site editor modal** 
- [x] Modal appears when editing or saving site modifications, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

**Post editor modal** 
- [x] Modal appears when updating or saving the post, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

#### Free Flow
**Site editor modal** 
- [x] Modal appears when editing or saving site modifications, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

**Post editor modal** 
- [x] Modal appears when updating or saving the post, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

#### Videopress
**Site editor modal** 
- [x] Modal appears when editing or saving site modifications, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

**Post editor modal** 
- [x] Modal appears when updating or saving the post, prompting the user to return the launchpad
- [x] Modal is keyboard navigable
- [x] "Don't show me again" setting ensures that the modal no longer appears on subsequent edits
- [x] Next steps button redirects user to launchpad 

### Browsers
- [x] Chrome

jeyip commented on projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js:
```diff
@@ -6,39 +6,54 @@ import { useSelect } from '@wordpress/data';
-		const isInsideSiteEditor = document.getElementById( 'site-editor' ) !== null;
+		const isInsideSiteOrBlockEditor =
+			document.getElementById( 'site-editor' ) !== null ||
+			document.querySelector( '.block-editor' ) !== null;
```
@mattwiebe Wondering if this is the best selector to use to identify the post editor 🤔

sharpirate replied:
I am open to suggestions here, I was a bit disappointed there was no unique ID I can use as there is for the site editor

jeyip commented on projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js:
```diff
@@ -6,39 +6,54 @@ import { useSelect } from '@wordpress/data';
+				 * It might have to do with the URL changing from /post-new.php to /post.php and potentially unmounting the component
+				 * A simple workaround I found to be working is to delay the rendering of the modal so the URL has already been changed by then
+				 */
+				setTimeout( () => {
```
Not a fan of using async code to address timing problems unless necessary. The order of code execution isn't guaranteed and it can lead to unexpected results that aren't intuitive to debug.

Is it possible to detect some prop that changes during the first save that we can use to determine when to render the modal? ( I'm fine with this being addressed in a follow-up )

sharpirate replied:
I agree, I am not a fan of this either but I wasn't able to find anything else. I think I can spend more time looking for sth in a follow up

jeyip commented on projects/plugins/jetpack/extensions/plugins/launchpad-save-modal/index.js:
```diff
@@ -6,39 +6,54 @@ import { useSelect } from '@wordpress/data';
-		const prevIsSaving = usePrevious( isSaving );
+		const isSavingPost = useSelect( select => select( editorStore ).isSavingPost(), [] );
+
+		const prevIsSavingSite = usePrevious( isSavingSite );
```
Would you mind elaborating on why we need the previous `isSaving` value?

sharpirate replied:
Because the selector is changing from `false` to `true` when you hit the publish button (when the saving is in progress) and back to `false`. I want to render the modal only after the post has been published (or the changes saved in the case of the site editor) -> That's why I check if the current value is `false` and the previous value was `true`. If I was to only check if the current value is `true` it would open the modal at the time of saving (still in progress)

jeherve commented:
>  I'm not seeing that:

Can you try typing some content, to let WordPress auto-save your draft post, and see if the modal comes up then?

https://user-images.githubusercontent.com/426388/213421319-a8573985-cb9a-4215-867d-8da84b9371fb.mov

nuriapenya commented:
Hello! Just to align:

> After a discussion inside cylon, we decided to have one generic version of the modal for all flows. I've also added support for the post editor. Feel free to review / test again.

Exactly. I think that having one generic message for all flows is the best course right now. When we look at the tailored messages, I feel a disconnection between the texts and the actions (both for Newsletter & LIB). Therefore, I'd say to go generic and default. And then, iterate in the future if we want to personalise them better.

Regarding:

> Should the modal appear even before the post is published? It seems to be happening right now, and I personally find it a bit annoying when I'm writing a draft.

We already have another modal in place - unless we replace one with the other. And bare in mind that the Loop Team is also working on improving that post-publish modal.
pe7F0s-bh-p2

<img width="1579" alt="Screenshot 2023-01-19 at 14 33 29" src="https://user-images.githubusercontent.com/7931304/213456546-e11a1e81-d061-4683-b6cb-e75a05c0fa76.png">

sharpirate commented:
FYI: @Automattic/cylon @jeherve I'll be closing this PR and I'll break it into two PRs - one for the site editor and another for the post editor (where I'll be addressing the above comments)

sharpirate commented:
[New Site Editor PR: Blocks: Make the Launchpad on save modal inside site editor generic for all flows](https://github.com/Automattic/jetpack/pull/28503)
